### PR TITLE
final for merge

### PR DESCRIPTION
**Pull request by gabyx**

### CODE REVIEW

**Comment by acki-m** [comment]
Honestly, my primary system is MSVC2017. But I care also for the other compilers.
I cannot merge your PR, because of this function `add_warning_levels`.
Your CMake macro is not my preferred, way, I would like to not touch the global CMake compiler flags anymore, when not necessary.
My approach with the `set_compiler_warnings` function seems better, so we are in full control
what target get what warnings. You are free to add your other warning flags to this functions.
We could also rename the function maybe to: 'set_default_compiler_warnings'.

So please remove this function.

File: .travis.yml [hunk 1]
```diff
@@ -38,7 +38,30 @@ env:
     - secure: "1ogFHnhUroT3nA1HxyXN5/Hoek6LcTazRPTGpYjKAge8To/x614D6R9j2yu0EDOZalolq0VeOONnHLnss6aPGeMvvHH1rvoKTDQsFjsm1fB6qLRr/qpZXs+G4mxStarxrXr5487hEkPP10fEi/PSOx8Gme6Kny+tXHxDzP04ljcZYVvxWfkb+cQBCnYupbHk4hzfKRmPYcUn6lCPyOGIzwsHQQI/LubmkiU+wQ0zcLLsTZSThk+zt6FlxYWykZGL24yrqmoXRszx8dR14cWPzvkxVeEOe3OY9mNix1Z7ki2eVeEk8m7d70yPmV5p8F2/nPGC7kUXTvPrAL62CcE5moHubijJwKXJM6nBQJYCPGygnvRnWxnoqO/9JpDuds4Cv+tbJMgo2ViLPsYk3BgMnYo4z+p4JViN9P4/jzbhsVtWd4rnXmPt9j4FkED+Eyt8BUzXFdTAPpuzwZSvmYzzR8MdZE/telCAcWUoiS35IQcWJ0UmPpNZfR1Lll0cWQ+dTmjNETJvwognuaTAexL7m0uMhX6KQIZfap2H0w+CIua7+jl6eb9iyvXv9vZZEUCH2kTWnYXqBmDGRHG3tEp97o5+yWQm1qjweLvKAMRirdkSaZEL0lrsJ8mBS6Y+PQwAQxlu+P3pFEJ6uU3fJrKyciuWnVDZKyddIrJEZorXjOA="
 
 matrix:
+  allow_failures:
+    - env: COMPILER=g++-7 ARCH=x86_64 CONF=Release UNIT_TESTS=true
   include:
```
Comment (acki-m) [inline .travis.yml]: remove

File: .travis.yml [hunk 2]
```diff
@@ -38,7 +38,30 @@ env:
...
+
+    # OSX/Clang
```
Comment (acki-m) [inline .travis.yml]: move MacOSX/ to last row again

File: CMake/utility.cmake [hunk 3]
```diff
@@ -386,6 +386,51 @@ function(activate_precompiled_headers _PRECOMPILED_HEADER _SOURCE_FILES)
   endif()
 endfunction()
 
+####################################################################################
+# Adds warnings compiler options to the target depending on the category
+# _TARGET Target name
+# _CATEGORY "library" or "unit_tests" or "benchmarks"
+####################################################################################
+macro( add_warning_levels _TARGET _CATEGORY)
```
Comment (acki-m) [inline CMake/utility.cmake]: remove function, see my `set_compiler_warnings`

File: src/rttr/detail/base/core_prerequisites.h [hunk 4]
```diff
@@ -62,6 +62,10 @@ namespace rttr
 // Compiler
 /////////////////////////////////////////////////////////////////////////////////////////
 #if defined( __clang__ )
+
+#if defined __apple_build_version__
```
Comment (acki-m) [inline src/rttr/detail/base/core_prerequisites.h:67]: Isn't the apple clang compiler because of its own version scheme,
a compiler for its own?
Because of that I would suggest, you tread him like an own compiler.
```
#if defined(__apple_build_version__)
#   define RTTR_COMPILER RTTR_COMPILER_APPLE_CLANG
```

Comment (gabyx) [reply src/rttr/detail/base/core_prerequisites.h:67]: jeah, I was think so too, but its still clang, an should behave like one, so there I decided, only to have a different version number, but jeah, maybe better to have it as an own compiler

File: .travis.yml (hunk 2 above)
Comment (gabyx) [reply .travis.yml]: jup

File: CMake/utility.cmake (hunk 3 above)
Comment (gabyx) [reply CMake/utility.cmake]: I ve seen, it: Can I had the "categories" too?, Such that I can switch for benchmarks?
I agree with you, that this function is only to add some warnings. Warning levels should not be touched, but since unit_tests and benchmarks are things on their own, its perfectly fine I think. "-Werror" is good and necessary such that other users have no problems using the library. If we only add some flags to make the benchmarks compile, this shouldnt be a problem right?

**Comment by gabyx** [comment]
@acki-m I think you understood wrong :-)
I never touched global compiler flags, never :-) > modern cmake is that way:
```
 target_compile_options(${_TARGET} PRIVATE ${CXX_FLAGS} )
  target_compile_options(${_TARGET} PRIVATE $<$<CONFIG:Debug>:${CXX_FLAGS_DEBUG}> )
  target_compile_options(${_TARGET} PRIVATE $<$<CONFIG:Release>:${CXX_FLAGS_RELEASE}> )
  target_compile_options(${_TARGET} PRIVATE $<$<CONFIG:MinSizeRel>:${CXX_FLAGS_MINSIZEREL}> )
  target_compile_options(${_TARGET} PRIVATE $<$<CONFIG:RelWithDebInfo>:${CXX_FLAGS_RELWITHDEBINFO}> )
```
and thats exactly what my function does already, yours and mine are basically the same :-)
its `CXX_FLAGS_DEBUG`  not `CMAKE_CXX_FLAGS_DEBUG`

**Comment by gabyx** [comment]
I thing we can test a final run where the warnings for the unit_tests are droppend, I experienced that it worked already, benchmarks dont, because of nonius...
try to drop:
`"-Wno-unused-private-field" "-Wno-delete-non-virtual-dtor"`
for unit_tests only

File: CMake/utility.cmake [hunk 5]
```diff
@@ -386,6 +386,51 @@ function(activate_precompiled_headers _PRECOMPILED_HEADER _SOURCE_FILES)
   endif()
 endfunction()
 
+####################################################################################
+# Adds warnings compiler options to the target depending on the category
+# _TARGET Target name
+# _CATEGORY "library" or "unit_tests" or "benchmarks"
+####################################################################################
+macro( set_compiler_warnings _TARGET _CATEGORY)
```
Comment (acki-m) [inline CMake/utility.cmake]: There are several things wrong with this function:
1. It is a macro, so you not introducing a new scope, the variables will be available where you call this macro. Prefer functions
2. DRY - e.g. you do this compare: `if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")` 3 times in this macro
3. You repeat yourself also with the warnings (e.g. "/WX" "/W4"), The whole code base should be compiled with the same warning levels.
When there are 3rd Party libs, which does not match this requirement, then we have to address this separately.
But int the current master state, all targets use the same compiler warnings. We can add later more warnings flags.
So in conclusion the category argument can be removed. Keep it simple.
4. Variable naming: `CXX_FLAGS` this are not a generic CXX flags, it should only be used for warnings, so I would call them like this.
5. `CXX_FLAGS_DEBUG` `CXX_FLAGS_RELEASE`, etc. Where are these coming from?

File: .travis.yml [hunk 6]
```diff
@@ -39,11 +39,7 @@ env:
 
 matrix:
   include:
-    # check code formatting
```
Comment (acki-m) [inline .travis.yml:42]: Why did you moved this line?
This should be the first check of the whole code base, formatting should be always checked first.

File: src/benchmarks/bench_method/CMakeLists.txt [hunk 7]
```diff
@@ -54,12 +54,11 @@ if(MSVC)
     target_compile_options(bench_method PRIVATE /Zm200)
 endif()
 
+set_compiler_warnings(bench_method "benchmarks")
```
Comment (acki-m) [inline src/benchmarks/bench_method/CMakeLists.txt]: I am not 100% sure, but I think adding the flags, before setting the CXX standard, will lead to effect that in verbose builds, we will see, which c++ standard is used, as last argument to the compiler. I prefer seeing it as first.

File: CMakeLists.txt [hunk 8]
```diff
@@ -35,6 +35,9 @@ cmake_minimum_required (VERSION 3.0)
 
 project ("rttr" LANGUAGES CXX)
 
+set(CMAKE_CXX_EXTENSIONS OFF)
```
Comment (acki-m) [inline CMakeLists.txt:38]: Why do you disable the CXX extensions?
This is the default behavior in CMake, I would stick to it.
Otherwise you can tell me a good reason why not.

**Review by acki-m — CHANGES_REQUESTED** [review]

**Comment by gabyx** [comment]
I disabled C++ Extensions because we don't need it and its not standard. I would prefere to stick to `-std=c++14` instead of using `-std=gnu++14` which is non-standard. When such a thing is needed in future, one can think about it again. CXX_EXTENSION_OFF is a reasonable choice, why cmake enables it be default I dont know? I think turning them off is good for a clean project as this one.

File: .travis.yml (hunk 6 above)
Comment (gabyx) [reply .travis.yml:42]: error

File: .travis.yml [hunk 9]
```diff
@@ -271,6 +275,7 @@ before_script:
       -DBUILD_WEBSITE_DOCU=$BUILD_DOCU
       -DBUILD_EXAMPLES=$BUILD_EXAMPLES
       -DBUILD_BENCHMARKS=$BUILD_BENCHMARKS
+      -DCMAKE_VERBOSE_MAKEFILE=ON
```
Comment (acki-m) [inline .travis.yml]: this should be removed, when its ready for merging

File: CMake/utility.cmake [hunk 10]
```diff
@@ -386,6 +386,25 @@ function(activate_precompiled_headers _PRECOMPILED_HEADER _SOURCE_FILES)
   endif()
 endfunction()
 
+####################################################################################
+# Adds warnings compiler options to the target depending on the category
+# _TARGET Target name
```
Comment (acki-m) [inline CMake/utility.cmake]: `_TARGET` wrong argument name, comment doesn't match

**Comment by acki-m** [comment]
@gabyx 
Okay we can leave it at first disabled (so we have the same behavior as before).
However, CMake has enabled it because, its enabled by the compilers by default.

**Comment by gabyx** [comment]
lets await the final merge travis builds (everything should be fine now) (except some travis builds which randomly dont build due to apt-get sometimes...)

**Comment by acki-m** [comment]
fixed in https://github.com/rttrorg/rttr/pull/100
